lib/sles4sap/sdaf_deployment_library.pm

```diff
@@ -1011,3 +1013,98 @@ sub sdaf_cleanup {
+
+B<env_code>: Code for SDAF deployment env. Default: 'SDAF_ENV_CODE'
+
+B<workload_vnet_code>: Virtual network code for workload zone. Default: 'SDAF_WORKLOAD_VNET_CODE'
+
+B<sdaf_region_code>: SDAF internal code for azure region. Default: 'PUBLIC_CLOUD_REGION' converted to SDAF format
+
+B<sap_sid>: SAP system ID. Default 'SAP_SID'
+
+B<timeout> Timeout for executing playbook. Passed into asset_script_run. Default: 1800s
+
+B<$verbosity_level> Change default verbosity value by either anything equal to 'true' or int between 1-6. Default: false
+
+Execute playbook specified by B<playbook_filename> and record command output in separate log file.
+Verbosity level of B<ansible-playbook> is controlled by openQA parameter B<SDAF_ANSIBLE_VERBOSITY_LEVEL>.
+If undefined, it will use standard output without adding any B<-v> flag. See function B<sdaf_execute_playbook> for details.
+
+=cut
+
+sub sdaf_execute_playbook {
+    my (%args) = @_;
+    $args{timeout} //= 1800;    # Most playbooks take more than default 90s
+    $args{env_code} //= get_required_var('SDAF_ENV_CODE');
+    $args{workload_vnet_code} //= get_required_var('SDAF_WORKLOAD_VNET_CODE');
+    $args{sdaf_region_code} //= convert_region_to_short(get_required_var('PUBLIC_CLOUD_REGION'));
+    $args{sap_sid} //= get_required_var('SAP_SID');
+    $args{verbosity_level} //= get_var('SDAF_ANSIBLE_VERBOSITY_LEVEL');
+
+    croak 'Missing mandatory argument "playbook_filename".' unless $args{playbook_filename};
+
+    my $sdaf_config_root_dir = get_sdaf_config_path(
+        deployment_type => 'sap_system',
+        vnet_code => $args{vnet_code},
+        env_code => $args{env_code},
+        sdaf_region_code => $args{sdaf_region_code},
+        sap_sid => $args{sap_sid}
+    );
+
+    my $playbook_options = join(' ',
+        sdaf_ansible_verbosity_level($args{verbosity_level}),    # verbosity controlled by OpenQA parameter
+        "--inventory-file=\"$args{sap_sid}_hosts.yaml\"",
+        "--private-key=$sdaf_config_root_dir/sshkey",
+        "--extra-vars='_workspace_directory=$sdaf_config_root_dir'",
+        '--extra-vars="@sap-parameters.yaml"',    # File is generated by SDAF, check official docs (SYNOPSIS) for more
+        '--ssh-common-args="-o StrictHostKeyChecking=no -o ServerAliveInterval=60 -o ServerAliveCountMax=120"');
+
+    assert_script_run("cd $sdaf_config_root_dir");
+    script_run("chmod 600 $sdaf_config_root_dir/sshkey");
+
+    my $output_log_file = log_dir() . "/$args{playbook_filename}" =~ s/.yaml|.yml/.log/r;
+    my $playbook_file = join('/', deployment_dir(), 'sap-automation', 'deploy', 'ansible', $args{playbook_filename});
+    my $playbook_cmd = join(' ', 'ansible-playbook', $playbook_options, $playbook_file);
+
+    record_info('Playbook run', "Executing playbook: $playbook_file\nExecuted command:\n$playbook_cmd");
+    my $rc = script_run(log_command_output(command => $playbook_cmd, log_file => $output_log_file),
```

Comment:
What is `log_command_output` ? O I see it is defined in `lib/sles4sap/sdaf_library.pm`

Reply:
Yes. It just adjusts the command string so the output is redirected to a file and returns correct RC.